### PR TITLE
wallet: fix crash during migration due to invalid multisig descriptors

### PR DESCRIPTION
**Pull request by furszy**
Ensure legacy wallet migration skips the never standard bare multisig with +3 keys
and consensus-invalid multisig scripts. Treating them as valid causes migration to
crash because we are enforcing this rules within the descriptors parsing logic.

Testing Notes:
This can be verified by cherry-picking and running the test commit on master.
It will crash there but pass on this branch.

### CODE REVIEW

**Review by darosior — COMMENTED** [review]
Ruling out consensus-invalid scripts makes sense, but discarding non-standard scripts is a much bigger leap. Also it would be a wack-a-mole: does the migration crash on any non-standard script? How much of that do you want to cover? Is it not possible to make the migration not crash on valid but not standard scripts instead of changing `IsMine` for those?

**Comment by furszy** [comment]
> Ruling out consensus-invalid scripts makes sense, but discarding non-standard scripts is a much bigger leap. Also it would be a wack-a-mole: does the migration crash on any non-standard script? How much of that do you want to cover?

Ok, I might have made the PR/commit description too broad when the actual changes are very specific.
Migration does not crash on all non-standard scripts, it only crashes on scripts that fail to be parsed by the descriptors' logic. The specific non-standard crash this PR addresses is a bare multisig with +3 keys. And I don't think we want to cover all non-standard rules; we just need to handle the ones also applied by the current descriptors parsing logic. 

Ultimately, we’re preserving/freezing the legacy wallet’s existing behavior ahead of its removal. We shouldn’t allow spendability or migration (even though it’s not possible anyway) of scripts that were not functional during the lifetime of the legacy wallet anyway.

> Is it not possible to make the migration not crash on valid but not standard scripts instead of changing `IsMine` for those?

It’s possible, but it would mean duplicating more code. Still, at least for this case, I don't think it matters much. #30328 inlines the `IsMine()` logic into the migration code, so the entire legacy wallet support can be removed by the next release.
In other words, duplicating code now won’t change the outcome; if #30328 proceeds, the code will end up been consolidated into the same migration function anyway.

I'll update the PR description to reflect the actual scope of this fix. it's not as broad as I initially described.

File: src/wallet/scriptpubkeyman.cpp [hunk 1]
```diff
@@ -203,6 +198,16 @@ IsMineResult IsMineInner(const LegacyDataSPKM& keystore, const CScript& scriptPu
...
             }
         }
+        // Follow consensus rules, never treat too large legacy multisig scripts as valid
+        if (sigversion == IsMineSigVersion::P2SH && scriptPubKey.size() > MAX_SCRIPT_ELEMENT_SIZE) {
+            return IsMineResult::INVALID;
+        }
+
+        // Never treat bare multisig outputs as ours (they can still be made watchonly-though)
+        if (sigversion == IsMineSigVersion::TOP) {
+            if (keys.size() > 3) return IsMineResult::INVALID; // These are standard wise non-spendable
```
Comment (brunoerg) [inline src/wallet/scriptpubkeyman.cpp:208]: nit: Maybe we should have a constant for this number (3)? Then we can use here and in `IsStandard`.

Comment (furszy) [reply src/wallet/scriptpubkeyman.cpp:208]: yeah, thats also something that popped-up in (https://github.com/bitcoin/bitcoin/pull/31404#discussion_r1866127089). Will bring that commit here too.

File: test/functional/wallet_migration.py [hunk 2]
```diff
@@ -299,6 +299,40 @@ def test_multisig(self):
...
+        res = wallet.createmultisig(4, pubkeys[:4])
+        # Import script as a bare multisig. This is standard-wise non-spendable, and it is not allowed descriptors' wise
+        wallet.importaddress(address=res['redeemScript'])
+
+        # Now migrate it and verify we don't crash due to a non-allowed descriptor migration
+        wallet.migratewallet()
+        wallet.unloadwallet()
+
+        ##############################################################
+        # Import a consensus-wise invalid p2sh multisig with 20 keys #
```
Comment (brunoerg) [inline test/functional/wallet_migration.py:324]: In d64193142792e68636bf61ef4b083dfb8c1d20d7: A p2sh multisig with 20 keys is invalid? I thought 20 keys are the limit, but still valid.

Comment (furszy) [reply test/functional/wallet_migration.py:324]: > In [d641931](https://github.com/bitcoin/bitcoin/commit/d64193142792e68636bf61ef4b083dfb8c1d20d7): A p2sh multisig with 20 keys is invalid? I thought 20 keys are the limit, but still valid.

Yes, they are. The limit is not based on the number of keys, it is on the size of the redeem script. Which can go up to 520 bytes max (which limits the number of keys inside a p2sh to 15).

**Comment by achow101** [comment]
cdaa3a58dc16d6a27248dc4cdec2dd1909eee7fe "wallet: bugfix, stop treating multisig consensus-invalid/unspendable scripts as ours" modifies legacy `IsMine()`, ostensibly to mix issues in migration. However, I don't think doing that makes sense that this point as legacy `IsMine` is going to be removed from migration with #30328, and any changes to it may result in incompatibilities with legacy wallets and migrated wallets.

I think that fixes to migration should be limited to the migration only code. Perhaps it would be easier to cherry pick the tests and migration specific fixes into #30328?

**Review by furszy — COMMENTED** [review]
> [cdaa3a5](https://github.com/bitcoin/bitcoin/commit/cdaa3a58dc16d6a27248dc4cdec2dd1909eee7fe) "wallet: bugfix, stop treating multisig consensus-invalid/unspendable scripts as ours" modifies legacy `IsMine()`, ostensibly to mix issues in migration. However, I don't think doing that makes sense that this point as legacy `IsMine` is going to be removed from migration with #30328, and any changes to it may result in incompatibilities with legacy wallets and migrated wallets.

I'm not going too strong over this point but.. it's hard for me to see the incompatibilities when the changes block two consensus-invalid scripts and a +10 year-old non-standard script that is already blocked in other parts of the legacy wallet. Arguably, this could also be labeled as a legacy wallet bugfix.

> I think that fixes to migration should be limited to the migration only code. Perhaps it would be easier to cherry pick the tests and migration specific fixes into https://github.com/bitcoin/bitcoin/pull/30328?

I was aiming to keep #30328 purely as a refactoring without mixing in bug fixes. Reviewing the PR's matching functionality has been challenging in its current state, so I think that it would be good if we don’t expand it more than strictly necessary.

Note: If, after this comment, the general consensus is still to leave `IsMine` untouched, maybe postponing this PR until after #30328 would be the path that satisfies everyone?

**Comment by achow101** [comment]
> Note: If, after this comment, the general consensus is still to leave `IsMine` untouched, maybe postponing this PR until after #30328 would be the path that satisfies everyone?

Then let's do it afterwards.

**Comment by Sjors** [comment]
Is this now pending #31495 instead of the closed #https://github.com/bitcoin/bitcoin/pull/30328?

**Comment by luke-jr** [comment]
> we are enforcing this rules within the descriptors parsing logic.

That sounds dumb. Why are we?

**Comment by pinheadmz** [comment]
@luke-jr can you rephrase your comment in a friendlier way please

**Comment by furszy** [comment]
> > we are enforcing this rules within the descriptors parsing logic.
> 
> That sounds dumb. Why are we?

I assume this was simply because there was no need to represent them in the descriptors' language. They've been non-standard for at least the past 10 years now.

However, there's no documentation about it. This restriction has always been present; it was introduced in the first commit that added descriptors in #13697.

We could relax the parsing restriction if a real use case arises. Until then, the descriptor's inference and parsing must allow for round-trips.

That being said, this PR might not end up getting merge. #31495 contains a general round-trip check that does not touch the legacy wallet `IsMine` code (haven't reviewed it in detail yet). And the fix for the descriptors issue is in #31404.

File: test/functional/wallet_migration.py [hunk 3]
```diff
@@ -299,6 +299,40 @@ def test_multisig(self):
...
         assert_equal(ms1_solvable.listtransactions(), [])
 
+    def test_multisig_invalid(self):
+        self.log.info("Test migration of a legacy-wise non-standard bare multisig")
+        wallet = self.create_legacy_wallet("multi_nonstandard")
+
+        # Create enough keys for all coming tests
+        addys = [wallet.getnewaddress()] * 20
+        pubkeys = []
+        privkeys = []
```
Comment (brunoerg) [inline test/functional/wallet_migration.py:309]: In 74fa29e12e379d7be8ad2dd261ee68efaf7a9e07: `privkeys` can be removed, it's unused.

**Comment by furszy** [comment]
closing for now. Not sure if I will ever back to work on this PR.
